File: sin_se.m
function [t]=sin_se(s)
%初始化
x = 1;
y = 0;
z = s * pi /180;
a = 0;
d = 1;
k = 0.6073; %k为增益
x = k*x;
while a<100 %此处不能判断d的负号控制循环，会死循环。应用a次数控制。
    if z>=0
        d = 1;
    else d = -1;
    end
%迭代
    xNew = x;
    x = xNew-(y*d*(1/2^a));
    y = y+(xNew*d*(1/2^a));
    z = z-(d*(atan(1/2^a)));
    a = a+1;
end
t = y;
end

File: sin_se.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define PI 3.141592

float sin_se(float);

float Factorial(int n)    //阶乘
{
	//long long 占8个字节
	float i, factorial = 1;

	for (i = 1; i <= n ; i++)
	{
		factorial *= i;
	}
	
	return factorial;
}

float sin_se(float x)
{
	float sin_se(x) = 0.0;
	int m;
	x = (fmod(x,360) * PI / 180);
	for(m = 1; m <= 12; m++)    //sinx的泰勒级数，m控制加到后面第几项
	{
		if (m % 2 == 0)
		{
			sin_se(x) -= pow(x, (2 * m - 1 )) / Factorial(2 * m - 1 );
		}
		else
		{
			sin_se(x) += pow(x, (2 * m - 1 )) / Factorial(2 * m - 1 );
		}
	}
	return sin_se(x);
}
